keel-bakery-plugin/src/main/kotlin/com/netflix/spinnaker/keel/bakery/artifact/BakeryLifecycleMonitor.kt

```diff
@@ -0,0 +1,120 @@
+import com.netflix.spinnaker.keel.lifecycle.LifecycleMonitorRepository
+import com.netflix.spinnaker.keel.lifecycle.MonitoredTask
+import com.netflix.spinnaker.keel.orca.ExecutionDetailResponse
+import com.netflix.spinnaker.keel.orca.OrcaExecutionStatus.BUFFERED
+import com.netflix.spinnaker.keel.orca.OrcaExecutionStatus.RUNNING
+import com.netflix.spinnaker.keel.orca.OrcaService
+import io.github.resilience4j.retry.Retry
+import io.github.resilience4j.retry.RetryConfig
+import io.vavr.control.Try
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.boot.context.properties.EnableConfigurationProperties
+import org.springframework.context.ApplicationEventPublisher
+import org.springframework.stereotype.Component
+import java.time.Duration
+
+@Component
+@EnableConfigurationProperties(LifecycleConfig::class)
+class BakeryLifecycleMonitor(
+  override val monitorRepository: LifecycleMonitorRepository,
+  override val publisher: ApplicationEventPublisher,
+  override val lifecycleConfig: LifecycleConfig,
+  private val orcaService: OrcaService,
+  @Value("\${spinnaker.baseUrl}") private val spinnakerBaseUrl: String
+): LifecycleMonitor(monitorRepository, publisher, lifecycleConfig) {
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  override fun handles(event: LifecycleEvent): Boolean =
+    event.type == BAKE
+
+  override fun typeHandled(): LifecycleEventType = BAKE
+
+  override suspend fun monitor(task: MonitoredTask) {
+    kotlin.runCatching {
+       orcaService.getOrchestrationExecution(task.link, DEFAULT_SERVICE_ACCOUNT)
+    }
+      .onSuccess { execution ->
+        when {
+          execution.status == BUFFERED -> publishCorrectLink(task)
+          execution.status == RUNNING -> publishRunningEvent(task)
+          execution.status.isSuccess() -> publishSucceededEvent(task)
+          execution.status.isFailure() -> publishFailedEvent(task)
+          else -> publishUnknownEvent(task, execution)
+        }
+
+        if (execution.status.isComplete()) {
+          endMonitoringOf(task)
+        }
+      }
+      .onFailure { exception ->
+        log.error("Error fetching status for $task: ", exception)
+        if (task.numFailures >= lifecycleConfig.numFailuresAllowed - 1) {
+          log.warn("Too many errors (${lifecycleConfig.numFailuresAllowed}) " +
+            "fetching the task status. Giving up.")
+          publishExceptionEvent(task)
+          monitorRepository.delete(task)
+        } else {
+          monitorRepository.markFailureGettingStatus(task)
+        }
+      }
+  }
+
+  private fun orcaTaskIdToLink(task: MonitoredTask): String =
+    "$spinnakerBaseUrl/#/tasks/${task.link}"
+
+  private fun publishCorrectLink(task: MonitoredTask) {
+    publisher.publishEvent(task.triggeringEvent.copy(
+      link = orcaTaskIdToLink(task)
+    ))
+  }
+
+  private fun publishRunningEvent(task: MonitoredTask) {
```

Comment:
consider extracting those functions to util class or something - I believe it will be useful in other places as well :)

Reply:
Can we do that when we need to? I think private functions / extensions like this are a good way to factor code. They can be moved to a common location once we find a need for them elsewhere. It doesn't seem likely to me that these would be used in another context.

Reply:
I don't want to because the thing we are actually doing is changing the text - which will be specific to each monitor.  Also, these are the tiniest baby functions. Feels ok to repeat them in other monitors (eventually) with other text.

As I build the build monitoring, though, I will see what can be extracted.